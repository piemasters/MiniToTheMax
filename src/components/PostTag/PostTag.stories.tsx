import type { Meta, StoryObj } from '@storybook/react-webpack5';

import { PostTag } from './PostTag';

const meta: Meta<typeof PostTag> = {
  component: PostTag,
  argTypes: {
    name: {
      name: 'name',
      description: 'The text within the tag',
      defaultValue: 'Tag',
      control: { type: 'text' },
    },
    type: {
      name: 'type',
      description: 'Used to determine the base url',
      defaultValue: 'tag',
      control: { type: 'text' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof PostTag>;

export const Primary: Story = {
  args: {
    name: 'Tag',
    type: 'tag',
  },
};
