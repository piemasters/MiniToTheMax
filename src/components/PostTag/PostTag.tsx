import React, { FC } from 'react';
import kebabCase from 'lodash.kebabcase';

import { PageLink } from '../PageLink';

export interface PostTagProps {
  name: string;
  type: string;
}

export const PostTag: FC<PostTagProps> = ({ name, type }) => {
  return (
    <div data-testid="post-tag" className="inline-block">
      <PageLink
        key={name}
        to={`/${type}/${kebabCase(name)}/`}
        className="bg-gray-200 rounded-md text-gray-800 cursor-pointer inline-block text-xs m-1 px-2 py-1 no-underline hover:bg-red-500 hover:text-white"
        type={'cover'}
        direction={'up'}
      >
        {name}
      </PageLink>
    </div>
  );
};

export default PostTag;
