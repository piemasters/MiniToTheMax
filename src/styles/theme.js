export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
