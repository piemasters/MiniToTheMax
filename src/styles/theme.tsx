export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
