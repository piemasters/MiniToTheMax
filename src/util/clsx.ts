/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
