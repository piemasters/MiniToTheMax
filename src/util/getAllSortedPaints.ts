import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
