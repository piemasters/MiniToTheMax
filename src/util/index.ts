export * from './pagination';
export * from './paintFilters';
