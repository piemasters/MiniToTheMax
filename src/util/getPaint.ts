import {
  CitadelPaintCategories,
  PaintCategory,
  PaintDetails,
  VallejoPaintCategories,
} from '../types';
import { getAllSortedPaints } from './getAllSortedPaints';

const allPaints = getAllSortedPaints();

const paintCategoryPriority = [
  ...CitadelPaintCategories,
  ...VallejoPaintCategories,
] as const;

type category = (typeof paintCategoryPriority)[number];

export const getPaint = (
  name: string,
  category?: PaintCategory
): { paint?: PaintDetails; error?: string; type?: string } => {
  const paintMatches = allPaints.filter((paint) => paint.name === name);

  if (paintMatches.length === 0) {
    return { error: 'Incorrect Paint Name', type: name };
  }

  if (paintMatches.length === 1) {
    return { paint: paintMatches[0] };
  }

  if (category) {
    const categoryMatches = paintMatches.filter(
      (paint) => paint.category === category
    );
    if (categoryMatches.length === 0) {
      return { error: 'Incorrect Paint Category', type: category };
    }
    return { paint: categoryMatches[0] };
  }

  const getCategoryPriority = (category: PaintCategory) => {
    const index = paintCategoryPriority.indexOf(category);
    return index === -1 ? Infinity : index;
  };
  paintMatches.sort(
    (a, b) => getCategoryPriority(a.category) - getCategoryPriority(b.category)
  );

  return { paint: paintMatches[0] };
};
