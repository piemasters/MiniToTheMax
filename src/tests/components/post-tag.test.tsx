import React from 'react';
import PostTag from '../../components/post-tag';
import { getByTestId } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import theme from '../../styles/theme';
import { renderWithTransitionProvider } from '../util/transition-provider';

describe('PostTag', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={theme}>
        <PostTag name={'Tag'} type={'tag'} />
      </ThemeProvider>
    );

    expect(getByTestId(container as HTMLElement, 'post-tag')).toHaveTextContent(
      'Tag'
    );
    expect(container).toMatchSnapshot();
  });
});
