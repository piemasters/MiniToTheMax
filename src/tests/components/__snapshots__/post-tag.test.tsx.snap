// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostTag renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="post-tag"
  >
    <a
      bg="#eb1d23"
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="page-link-cover"
      direction="up"
      duration="1"
      href="/tag/tag/"
    >
      Tag
    </a>
  </div>
</div>
`;
